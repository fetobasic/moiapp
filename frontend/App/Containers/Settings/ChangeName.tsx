import React, { useState } from 'react';
import { View, KeyboardAvoidingView, Keyboard, Text, StyleSheet } from 'react-native';
import { NativeStackScreenProps } from '@react-navigation/native-stack';

import { Button, HeaderSimple as Header, TextInput } from 'App/Components';

import { ApplicationStyles, Colors, Metrics, isIOS } from 'App/Themes';
import { SettingsStackParamList } from 'App/Types/NavigationStackParamList';
import { useAppDispatch } from 'App/Hooks';
import { showInfo } from 'App/Services/Alert';
import { devicesActions } from 'App/Store/Devices';
import { isLegacyYeti } from 'App/Services/Yeti';
import { update } from 'App/Services/ConnectionControler';
import ValidationErrorList from 'App/Components/ValidationErrorList';

type Props = NativeStackScreenProps<SettingsStackParamList, 'ChangeName'>;

function ChangeName({ route, navigation }: Props) {
  const { device } = route.params;
  const dispatch = useAppDispatch();

  const [name, setName] = useState(device?.name || '');
  const [isLoading, setIsLoading] = useState(false);
  const isBtnDisabled = name === device.name;

  const handleChangeName = (skipModal?: boolean) => {
    // Check for empty name
    if (/^\s*$/.test(name)) {
      showInfo("Name can't be empty");
      return;
    }

    const showModal = () =>
      showInfo(
        <Text>
          Your Goal Zero device was renamed from “<Text style={styles.name}>{device.name}</Text>” to "
          <Text style={styles.name}>{name}</Text>".
        </Text>,
        'Renamed!',
        () => navigation.goBack(),
      );

    Keyboard.dismiss();
    setIsLoading(true);

    const desired = isLegacyYeti(device.thingName || '') ? { name } : { identity: { lbl: name } };

    dispatch(
      devicesActions.devicesChangeName({
        thingName: device.thingName || '',
        name,
        cb: () => {
          setIsLoading(false);

          if (skipModal) {
            navigation.goBack();
          } else {
            showModal();
          }
        },
      }),
    );

    // send request to Direct or AWS API
    // @ts-ignore TODO: `update` function signature must be refactored with complex and recursive Partial stateObject
    update({
      thingName: device.thingName || '',
      method: 'device',
      stateObject: { state: { desired } },
      isDirectConnection: Boolean(device.isDirectConnection),
    });
  };

  return (
    <View style={ApplicationStyles.mainContainer}>
      <Header title="Name" isChangeSaved={!isBtnDisabled} cbSave={() => handleChangeName(true)} />
      <KeyboardAvoidingView style={styles.container} behavior={isIOS ? 'padding' : 'height'}>
        <View style={[ApplicationStyles.section, styles.container]}>
          <TextInput
            focused
            placeholder="Device Name"
            value={name}
            onChangeText={setName}
            returnKeyType="go"
            maxLength={32}
          />
          <ValidationErrorList style={styles.validation} errorList={['Max 32 Characters']} />
        </View>
        <Button
          testID="saveBtn"
          style={styles.saveBtn}
          title="SAVE"
          showLoading={isLoading}
          onPress={handleChangeName}
          disabled={isBtnDisabled}
          inverse={isBtnDisabled}
        />
      </KeyboardAvoidingView>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  name: {
    color: Colors.green,
  },
  validation: {
    paddingHorizontal: Metrics.halfMargin,
    marginBottom: Metrics.marginSection,
  },
  saveBtn: {
    marginHorizontal: Metrics.marginSection,
    marginVertical: Metrics.bigMargin,
  },
});

export default ChangeName;
