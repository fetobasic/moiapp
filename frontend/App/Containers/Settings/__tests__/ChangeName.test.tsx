import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { mockedState } from 'App/Fixtures/mocks/mockedState';
import { yeti6GState } from 'App/Fixtures/yeti6GState';
import ChangeName from '../ChangeName';

const mockStore = createMockStore([]);
const store = mockStore({
  ...mockedState,
});

describe('ChangeName', () => {
  test('renders screen', () => {
    const props = {
      navigation: {
        navigate: jest.fn() as any,
        canGoBack: jest.fn() as any,
      },
      route: { params: { device: yeti6GState } },
    } as any;

    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <ChangeName {...props} />
        </NavigationContainer>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('renders screen press button', () => {
    const props = {
      navigation: {
        navigate: jest.fn() as any,
        canGoBack: jest.fn() as any,
      },
      route: { params: { device: yeti6GState } },
    } as any;

    const { getByTestId } = render(
      <Provider store={store}>
        <NavigationContainer>
          <ChangeName {...props} />
        </NavigationContainer>
      </Provider>,
    );

    const saveBtn = getByTestId('saveBtn');

    fireEvent.press(saveBtn);
  });
});
