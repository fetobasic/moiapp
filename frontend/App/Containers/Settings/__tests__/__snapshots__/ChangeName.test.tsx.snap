// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeName renders screen 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(27, 27, 27, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(27, 27, 27, 1)",
          "borderBottomLeftRadius": 24,
          "borderBottomRightRadius": 24,
          "elevation": 12,
          "shadowColor": "rgba(0, 0, 0, 1)",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 12,
          "zIndex": 1,
        },
        {
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "justifyContent": "center",
          "marginHorizontal": 24,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={20}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            {
              "left": 0,
              "padding": 10,
              "position": "absolute",
            },
          ]
        }
      >
        <SvgMock />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 48,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RobotoCondensed-Medium",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 32,
            }
          }
        >
          Name
        </Text>
      </View>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "marginHorizontal": 24,
            "paddingTop": 16,
          },
          {
            "flex": 1,
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(27, 27, 27, 1)",
            "elevation": 12,
            "height": 56,
            "shadowColor": "rgba(0, 0, 0, 1)",
            "shadowOffset": {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 0.2,
            "shadowRadius": 12,
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "borderColor": "rgba(54, 54, 54, 1)",
              "borderRadius": 8,
              "borderWidth": 1,
              "flex": 1,
              "height": 56,
              "justifyContent": "center",
            }
          }
        >
          <Text
            collapsable={false}
            numberOfLines={1}
            style={
              {
                "backgroundColor": "rgba(27, 27, 27, 1)",
                "color": "rgba(255, 255, 255, 0.871)",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 20,
                "marginLeft": 8,
                "paddingHorizontal": 4,
                "position": "absolute",
                "top": 16,
              }
            }
          >
            Device Name
          </Text>
          <TextInput
            autoFocus={true}
            focused={true}
            keyboardAppearance="dark"
            maxFontSizeMultiplier={1.15}
            maxLength={32}
            numberOfLines={1}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            onPressIn={[Function]}
            placeholder=""
            placeholderTextColor="rgba(255, 255, 255, 0.38)"
            returnKeyType="go"
            selectionColor="rgba(191, 210, 43, 1)"
            style={
              [
                {
                  "color": "rgba(255, 255, 255, 0.87)",
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "lineHeight": 16,
                  "paddingHorizontal": 12,
                },
                undefined,
              ]
            }
            underlineColorAndroid="rgba(0,0,0,0)"
            value="gzy1-test"
          />
        </View>
      </View>
      <View
        style={
          {
            "marginBottom": 16,
            "paddingHorizontal": 12,
          }
        }
      >
        <Text
          style={
            {
              "color": "rgba(255, 255, 255, 0.87)",
              "fontFamily": "Roboto-Medium",
              "fontSize": 12,
              "fontWeight": "300",
              "lineHeight": 16,
              "marginTop": 8,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgba(255, 83, 29, 1)",
              }
            }
          >
            *
          </Text>
           
          Max 32 Characters
        </Text>
      </View>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": true,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "opacity": 1,
          },
          [
            {
              "elevation": 12,
              "shadowColor": "rgba(0, 0, 0, 1)",
              "shadowOffset": {
                "height": 4,
                "width": 4,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 12,
            },
            {
              "marginHorizontal": 16,
              "marginVertical": 32,
            },
          ],
        ]
      }
      testID="saveBtn"
    >
      <View
        style={
          {
            "borderRadius": 24,
            "overflow": "hidden",
          }
        }
      >
        <RNSVGSvgView
          bbHeight="48"
          bbWidth="100%"
          focusable={false}
          height={48}
          style={
            [
              {
                "backgroundColor": "transparent",
                "borderWidth": 0,
              },
              {
                "flex": 0,
                "height": 48,
                "width": "100%",
              },
            ]
          }
          width="100%"
        >
          <RNSVGGroup
            fill={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
          >
            <RNSVGDefs>
              <RNSVGLinearGradient
                gradient={
                  [
                    0,
                    -13224394,
                    1,
                    1776411,
                  ]
                }
                gradientTransform={null}
                gradientUnits={0}
                name="grad"
                x1="0"
                x2="0"
                y1="0"
                y2="0.5"
              />
            </RNSVGDefs>
            <RNSVGRect
              fill={
                {
                  "brushRef": "grad",
                  "type": 1,
                }
              }
              height="48"
              propList={
                [
                  "fill",
                ]
              }
              width="100%"
              x="0"
              y="0"
            />
            <View
              style={
                [
                  {
                    "borderRadius": 24,
                    "marginHorizontal": 1,
                    "marginTop": 1,
                  },
                  {
                    "backgroundColor": "rgba(27, 27, 27, 1)",
                    "height": 46,
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                      "paddingVertical": 12,
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "rgba(255, 255, 255, 1)",
                          "fontFamily": "Roboto-Medium",
                          "fontSize": 14,
                          "fontWeight": "700",
                          "lineHeight": 20,
                          "textTransform": "uppercase",
                        },
                        {
                          "color": "rgba(255, 255, 255, 0.38)",
                        },
                        undefined,
                      ]
                    }
                  >
                    SAVE
                  </Text>
                </View>
              </View>
            </View>
          </RNSVGGroup>
        </RNSVGSvgView>
      </View>
    </View>
  </View>
</View>
`;
